File: 104.Maximum_Depth_of_Binary_Tree.cc
#include <iostream>
#include <cassert>

class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (!root) return 0;
        return (std::max(maxDepth(root->left), maxDepth(root->right)) + 1);
    }
};

/*
    0
   / \
  1   2

  std::max(root->left, root->right)  on 1 results in 0 since it has no children, add 1 and returns up
  node 0 now has a max of (1,1) to pick from which is 1, add 1 for its own level 
  returns 2
*/

int main() {
    Solution solution;

    assert(solution);

    std::cout << "\033[32mAll tests passed!\033[0m\n";
    return 0;
}

File: Python solutions/104. Maximum Depth of Binary Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root: return 0

        def dfs(node: Optional[TreeNode]) -> int:
            if not node: return 0
            return max(dfs(node.left), dfs(node.right)) + 1

        return dfs(root)
        
assert Solution(). == , "Test failed!"

print("\033[32mAll tests passed!\033[0m\n")